autoload/lsp/ui/vim/utils.vim

```diff
@@ -9,34 +9,69 @@ function! lsp#ui#vim#utils#locations_to_loc_list(result) abort
 
     if !empty(l:locations) " some servers also return null so check to make sure it isn't empty
         let l:cache={}
-        for l:location in l:locations
-            if s:is_file_uri(l:location['uri'])
-                let l:path = lsp#utils#uri_to_path(l:location['uri'])
-                let l:line = l:location['range']['start']['line'] + 1
-                let l:char = l:location['range']['start']['character']
-                let l:col = lsp#utils#to_col(l:path, l:line, l:char)
-                let l:index = l:line - 1
-
-                if has_key(l:cache, l:path)
-                    let l:text = l:cache[l:path][l:index]
-                else
-                    let l:contents = getbufline(l:path, 1, '$')
-                    if !empty(l:contents)
-                        let l:text = l:contents[l:index]
+        if has_key(l:locations[0],'uri') " server returns Locations
```

Comment:
That would pretty much only be the caching mechanism, right? What would be the best signature? 
`s:cache_text(l:cache,l:path,l:line)`, returning `l:text`?

Alternatively, the `locationLink` codepath could be changed to read only the actually needed lines, avoiding unnecessary work?

Reply:
The only difference between the cases are these two lines, I believe?
```vim
let l:viewstart = l:location['targetRange']['start']['line'] 
let l:viewend = l:location['targetRange']['end']['line']
```

Reply:
and the `uri`/`targetUri` in the two lines before

Reply:
I think I'd rather have the `if` just around the things that are different (more deeply nested, I guess).

Reply:
Hmm, I was debating that. I thought it'd be better to have a single check outside of the loop instead of repeatedly doing (possibly very often) the same check with guaranteed same result...

But I can change that if you're sure it doesn't influence performance.

Reply:
Well, Knuth once said that "premature optimisation is the root of all evil" :)
I'd rather lose some processor cycles, than have duplication, because in the long term, that will make maintenance difficult.
If you're worried about performance, you can store the result of the check in a variable:
```vim
let l:use_location_links = ...
```
and then use that in the rest of the function:
```vim
let l:line = l:use_location_links ? ... : ...
...
```

Reply:
Fair enough. But (not wishing to debate _too much_), the only common line beside the caching is really only 
```vimL 
let l:col = lsp#utils#to_col(l:path, l:line, l:char)  
```
so just refactoring the middle caching part seems cleaner.

Alternatively, I could try to extract the differing strings into a common `l:uri` and `l:range`, so the check would only be needed once (still in each loop)?

Reply:
`l:uri` and `l:range` sounds good. Then you would only need an `if` around that and the call to `add`.

Sorry if I exaggerate a bit with my reviews, it's a fault of mine. Feel feel to let me know when my nitpicking becomes too much :)

Reply:
No worries, clean code is happy code!

Feel free to bikeshed on the fields and names (and let me know if something should be documented better).